Code/GraphMol/FileParsers/file_parsers_catch.cpp

```diff
@@ -6928,4 +6932,124 @@ TEST_CASE(
     REQUIRE(nm);
     CHECK(MolToSmarts(*nm) == "[#6,#7,#8][#6][#6]*[$(C(=O)O)]");
   }
-}
\ No newline at end of file
+}
+
+class FragTest {
+ public:
+  std::string fileName;
+  bool expectedResult;
+  bool reapplyMolBlockWedging;
+
+  FragTest(std::string fileNameInit, bool expectedResultInit,
+           bool reapplyMolBlockWedgingInit = true)
+      : fileName(fileNameInit),
+        expectedResult(expectedResultInit),
+        reapplyMolBlockWedging(reapplyMolBlockWedgingInit){};
+};
+
+void generateNewExpectedFilesIfSoSpecified(std::string filename,
+                                           std::string dataToWrite) {
+  if (generateExpectedFilesFlag) {
+    std::ofstream out;
+    out.open(filename);
+    out << dataToWrite;
+  }
+}
+
+std::string GetExpectedValue(std::string expectedFileName) {
+  std::stringstream expectedMolStr;
+  std::ifstream in;
+  in.open(expectedFileName);
+  expectedMolStr << in.rdbuf();
+  return expectedMolStr.str();
+}
+
+void testFragmetation(const FragTest &fragTest) {
+  INFO(fragTest.fileName);
+  std::string rdbase = getenv("RDBASE");
+
+  std::string fName = rdbase +
+                      "/Code/GraphMol/FileParsers/test_data/sgroupFragments/" +
+                      fragTest.fileName;
+  std::unique_ptr<RWMol> mol(MolFileToMol(fName, false));  // don't sanitize yet
+  REQUIRE(mol);
+
+  auto frags = MolOps::getMolFrags(*mol, true);
+  CHECK(frags.size() > 1);
+
+  // get the largest fragment
+
+  unsigned int largestFragSize = 0;
+  ROMol *largestFrag = nullptr;
+  for (auto frag : frags) {
+    if (frag->getNumAtoms() > largestFragSize) {
+      largestFragSize = frag->getNumAtoms();
+      largestFrag = frag.get();
+    }
+  }
+
+  CHECK(largestFrag);
+
+  {
+    std::string expectedName =
+        fName + (fragTest.reapplyMolBlockWedging ? "" : ".noReapply") +
+        ".expected.sdf";
+    std::string outMolStr = MolToMolBlock(*largestFrag, true, 0, true, true);
+
+    generateNewExpectedFilesIfSoSpecified(
+        fName + (fragTest.reapplyMolBlockWedging ? "" : ".noReapply") +
+            ".NEW.sdf",
+        outMolStr);
+
+    TEST_ASSERT(GetExpectedValue(expectedName) == outMolStr);
+  }
+}
+
+TEST_CASE("FragmentSgroupTest", "[bug][reader]") {
```

Comment:
I'm not a big fan of the style of testing where there's a mechanical comparison between the generated output and some expected output except for in cases where you're testing the ability to actually generate the output.

In this case the test is to make sure that the SGroups correctly survive fragmentation, so I think it makes sense to have tests more of the form:
 - read molecule with S groups satisfying a particular condition
 - verify that condition on the input molecule, partially as documentation, partially to be sure that you're actually starting from the correct place
 - transform the molecule (in this case fragment it)
 - explicitly verify that the expected S groups are still there

With the current implementation of the fix, this would require a lot of molecules (as you have), but with a simplified fix - coming in a PR against your branch, this would no longer be necessary.

Reply:
I have merged your changes and reworked the tests as suggested.

Reply:
Thanks